sources/settings-migrations/v1.40.0/kubelet-device-plugins-cdi-settings/src/main.rs

```diff
@@ -0,0 +1,216 @@
+use migration_helpers::{error, migrate, Migration, MigrationData, Result};
+use snafu::OptionExt;
+
+const DEVICE_LIST_STRATEGY_SETTING: &str =
+    "settings.kubelet-device-plugins.nvidia.device-list-strategy";
+
+#[snafu::report]
+fn main() -> Result<()> {
+    migrate(ReplaceDeviceListStrategy)
+}
+
+/// We changed the type of the device-list-strategy in the NVIDIA Kubernetes Device
+/// plugin API from a string to a list, and accept "cdi-cri" as a valid value
+pub struct ReplaceDeviceListStrategy;
+
+impl Migration for ReplaceDeviceListStrategy {
+    /// New versions must either have a default for the settings or generate them; we don't need to
+    /// do anything.
+    fn forward(&mut self, input: MigrationData) -> Result<MigrationData> {
+        println!("ReplaceDeviceListStrategy has no work to do on upgrade.",);
+        Ok(input)
+    }
+
+    /// Older versions don't know about the settings; we remove them so that old versions don't see
+    /// them and fail deserialization.  (The settings must be defaulted or generated in new versions,
+    /// and safe to remove.)
+    fn backward(&mut self, mut input: MigrationData) -> Result<MigrationData> {
+        let setting = DEVICE_LIST_STRATEGY_SETTING;
+        if let Some(data) = input.data.get_mut(setting) {
+            match data {
+                serde_json::Value::Array(arr) => {
+                    let list: Vec<&str> = arr
+                        .iter()
+                        .map(|v| v.as_str())
+                        .collect::<Option<Vec<&str>>>()
+                        .context(error::ReplaceListContentsSnafu {
+                            setting,
+                            data: arr.clone(),
+                        })?;
+
+                    let new_value = if list.is_empty() || list[0] == "cdi-cri" {
+                        "volume-mounts".to_string()
+                    } else {
+                        list[0].to_string()
+                    };
+
+                    *data = serde_json::Value::String(new_value.to_string());
+                }
+
+                serde_json::Value::String(setting_str) => {
+                    if setting_str == "cdi-cri" {
+                        *data = serde_json::Value::String("volume-mounts".to_string());
+                    } else {
+                        *data = serde_json::Value::String(setting_str.to_string());
+                    }
```

Comment:
Given that the setting values are basically enum, can this be:

```suggestion
                   *data = serde_json::Value::String(setting_str.replace("cdi-cri", "volume-mounts").to_string());
```

Reply:
I would like to keep the if/else since the logic is more user readable.